Interprocedural propagation of slot refinements

serenity4 opened the pull request:
This PR aims to propagate slot refinement information interprocedurally, which allows inference to narrow down the type of a slot as follows:

```julia
subfunc(x) = x::Float64 + 1
function topfunc(x) # `topfunc` will be inferred with `x::Any`
    y = subfunc(x)
    return sin(x), cos(y)
end
src = code_typed1(topfunc, (Any,); optimize = false)
```

Before:
```julia
CodeInfo(
1 ─ %1 = Main.subfunc::Core.Const(subfunc)
│        (y = (%1)(x))::Float64
│   %3 = Main.sin::Core.Const(sin)
│   %4 =   dynamic (%3)(x)::Any
│   %5 = Main.cos::Core.Const(cos)
│   %6 = y::Float64
│   %7 =   dynamic (%5)(%6)::Float64
│   %8 =   builtin Core.tuple(%4, %7)::Tuple{Any, Float64}
└──      return %8
)
```

After:
```julia
CodeInfo(
1 ─ %1 = Main.subfunc::Core.Const(subfunc)
│        (y = (%1)(x))::Float64
│   %3 = Main.sin::Core.Const(sin)
│   %4 =   dynamic (%3)(x)::Float64
│   %5 = Main.cos::Core.Const(cos)
│   %6 = y::Float64
│   %7 =   dynamic (%5)(%6)::Float64
│   %8 =   builtin Core.tuple(%4, %7)::Tuple{Float64, Float64}
└──      return %8
)
```

The general mechanism is a form of reverse type inference (in the sense that it looks at how the program uses a value to deduce its type), providing an upper bound on types which inference may then use within control-flow paths assumed to be error-free. For example, it should eventually enable accurate inference for code such as

```julia
f1(x::Number) = x
f2(x::Signed) = x
f2(x::Symbol) = x
f3(x::Integer) = x::Int64
f3(x::AbstractFloat) = x::Float64

g1(x) = x::AbstractString
g2(x) = isa(x, String) ? x : throw(ArgumentError("Expected a String"))

function f(x, cond::Bool) # assume `f` will be inferred with `x::Any`
    if cond
        f1(x) # implies x::Number (assuming a method match)
        f2(x) # implies x::Signed (assuming a method match)
        f3(x) # implies x::Integer -> x::Int64 (assuming a method match and a non-throwing type assert)
    else
        g1(x) # implies x::AbstractString (assuming a method match and a non-throwing type assert)
        g2(x) # implies x::String (assuming a method match and a non-throwing execution)
    end
    x # inferred to x::Union{Int64, String}
end

code_typed(f, (Any, Bool); optimize = false)
```

A local version was already been implemented for slots with a slot refinements mechanism, where e.g. type asserts and method matches[^1] *within the same function* are used to tighten slot types (see https://github.com/JuliaLang/julia/issues/55229). In addition to being limited to slots, the current implementation only propagates the information forward from the point where it is obtained. While future work may attempt to lift these limitations, the intention for this PR is only to propagate this information interprocedurally, so that refinements for slots within a callee may be mapped back to slot refinements within the caller.

We may distinguish between two types of refinement:
- Local refinement, which contains type assumptions made along a program path (e.g. based on `isa` branches or type assertions made in a given block).
- Global refinement, which contains assumptions that hold regardless of control-flow.

Global refinement is the only information that can be reliably propagated across function boundaries, allowing a caller to tighten the type of arguments provided to a callee when making the assumption that the call does not throw. It may be derived from local refinements using a merging mechanism across control-flow paths similarly to return type inference. As an example, `isa` conditionals will globally tighten the type of their input only if they are used to branch and one of the branches throws.

Note that because the core assumption for this type refinement mechanism is that the execution does not throw, this information would have to be discarded at `try` block boundaries.

[^1]: Relying on a finite set of method matches to deduce type information about their potential arguments may introduce more sources of invalidation.

serenity4 committed:
Implement very basic interprocedural slot refinement

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Fixes

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Misc fixes & cleanup

topolarity commented:
> reverse type inference

I've always been a bit confused by this - why is it reverse? I thought the more-precise type-information only applies from the type-assert onward

serenity4 commented on Compiler/src/abstractinterpretation.jl:
```diff
@@ -588,11 +605,25 @@ function collect_slot_refinements(𝕃ᵢ::AbstractLattice, applicable::Vector{M
-            if sigt ⊏ argt # i.e. signature type is strictly more specific than the type of the argument slot
+            j = findfirst(x -> x.slot === fargᵢ, refinements)
+            newt = j === nothing ? sigt : begin
+                inferred = refinements[j].typ
+                sigt ⊑ inferred ? sigt : inferred ⊑ sigt ? inferred : begin
+                    # Core.println("sigt and inferred are not ordered: ", sigt, ", ", inferred)
+                    sigt
+                end
```
This is where refinement information obtained from callees is hooked into the slot refinement mechanism. Setting this to `sigt` restores the previous behavior by discarding interprocedural information.

Unfortunately at the moment it seems that enabling this integration gives issues to bootstrapping, with
```
Compiling the compiler. This may take several minutes ...
error during bootstrap:
LoadError(at "Base_compiler.jl" line 3: BoundsError(a=Array{TypeVar, 1}(dims=(1,), mem=Memory{TypeVar}(8, 0x7f31788dbf80)[
  𝕃ᵢ,
  #<null>,
  #<null>,
  #<null>,
  #<null>,
  #<null>,
  #<null>,
  #<null>]), i=(0,)))
unknown function (ip: 0x7f317b6c48f9) at (unknown file)
unknown function (ip: 0x7f317b6c78da) at (unknown file)
unknown function (ip: 0x7f317b6c9aa8) at (unknown file)
unknown function (ip: 0x7f317b6ca5b9) at (unknown file)
_jl_invoke at /home/serenity4/julia/wip3/src/gf.c:3500 [inlined]
ijl_apply_generic at /home/serenity4/julia/wip3/src/gf.c:3700
unknown function (ip: 0x7f31841a020d) at (unknown file)
_jl_invoke at /home/serenity4/julia/wip3/src/gf.c:3500 [inlined]
```

although setting this to `sigt` then revising changes after the build shows no errors while running the test suite.

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Correctly merge slot refinements across method matches

serenity4 committed:
Use `Core.Compiler.return_type`

serenity4 committed:
Improve bail out logic when callee != called

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Attempt to work around lack of tmeet symmetry

serenity4 committed:
Merge slot refinements across control-flow paths

serenity4 committed:
Fixes for control-flow handling

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Support enabling/disabling IPO slot refinement via InferenceParams

serenity4 committed:
Don't use StateUpdate changes for IPO refinements

serenity4 committed:
Only record refinements for argument slots

Non-argument slots should not be of relevance for IPO,
and the local refinement mechanism via `currstate::VarTable`
should already apply all refinements within the inference frame.

serenity4 committed:
Consider NOT_FOUND to be equivalent to Bottom post-inference

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Add missing conditional

serenity4 committed:
Remove assertion on argument slot reassignment

serenity4 committed:
Work around `tmeet` limitation

serenity4 committed:
Re-enable IPO slot propagation by default

serenity4 committed:
Add basic logging & filtering for sequential Compiler tests

serenity4 commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

serenity4 committed:
Don't construct `SlotRefinementPropagation` if disabled

serenity4 committed:
Comment out slot refinement IPO to troubleshoot performance

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Implement partial `tmeet` situational workaround

serenity4 committed:
Enable slot refinement IPO by default

serenity4 committed:
Disable slot refinement IPO again to investigate possible regressions

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

serenity4 committed:
Move InferenceState constructor outside of constructor

This is to have Revise work well with it when the
constructor gets modified, as moving this method
doesn't seem to have any disadvantages

serenity4 committed:
Optimize slot refinement IPO a bit

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/57651#issuecomment-3064492583) failed.

serenity4 committed:
Misc fixes

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/57651#issuecomment-3066345123) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6c9a80d_vs_6ddb3d6/report.md).

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into propagate-vartable

serenity4 committed:
Revert making InferenceState constructor an outer one

serenity4 committed:
Remove some noise from the changes
